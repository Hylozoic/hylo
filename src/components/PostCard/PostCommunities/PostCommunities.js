import React, { Component } from 'react'
import { get, isEmpty } from 'lodash/fp'
import cx from 'classnames'
import { Link } from 'react-router-dom'
import { communityUrl } from 'util/navigation'
import CommunitiesList from 'components/CommunitiesList'
import Icon from 'components/Icon'
import './PostCommunities.scss'

export default class PostCommunities extends Component {
  static defaultState = {
    expanded: false
  }

  constructor (props) {
    super(props)
    this.state = PostCommunities.defaultState
  }

  toggleExpanded = () => {
    this.setState({
      expanded: !this.state.expanded
    })
  }

  render () {
    const { communities, constrained, slug, showBottomBorder } = this.props
    const { expanded } = this.state

    // don't show if there are no communities or this isn't cross posted
    if (isEmpty(communities) || (communities.length === 1 && get('0.slug', communities) === slug)) return null

    return <div styleName={cx('communities', { constrained, expanded, bottomBorder: showBottomBorder })} onClick={expanded ? this.toggleExpanded : undefined}>
      <div styleName='row'>
        <span styleName='label'>Posted In:&nbsp;</span>
        {!expanded &&
          <LinkedCommunityNameList communities={communities} maxShown={2} expandFunc={this.toggleExpanded} />}
        <a onClick={this.toggleExpanded} styleName='expandLink'><Icon name={expanded ? 'ArrowUp' : 'ArrowDown'} styleName='expandIcon' /></a>
      </div>

      {expanded && <CommunitiesList communities={communities} />}
    </div>
  }
}

export function LinkedCommunityNameList ({ communities, maxShown = 2, expandFunc }) {
  const communitiesToDisplay = (maxShown && maxShown <= communities.length)
    ? communities.slice(0, maxShown)
    : communities
  const othersCount = communities.length - communitiesToDisplay.length

  return <span styleName='communityList'>
    {communitiesToDisplay.map((community, i) =>
      <LinkedCommunityName community={community} key={i}>
        <Separator currentIndex={i} displayCount={communitiesToDisplay.length} othersCount={othersCount} />
      </LinkedCommunityName>)}
    {othersCount > 0 &&
      <Others othersCount={othersCount} expandFunc={expandFunc} />}
  </span>
}

export function LinkedCommunityName ({ community, children }) {
  return <span key={community.id}>
    <Link to={communityUrl(community.slug)} styleName='communityLink'>{community.name === 'Public' && <Icon name='Public' styleName='publicCommunityIcon' />} {community.name}</Link>
    {children}
  </span>
}

export function Separator ({ currentIndex, displayCount, othersCount }) {
  const isLastEntry = currentIndex === displayCount - 1
  const isNextToLastEntry = currentIndex === Math.max(0, displayCount - 2)
  const hasOthers = othersCount > 0

  if (isLastEntry) return null
  if (!hasOthers && isNextToLastEntry) return <span key='and'> and </span>

  return <span>, </span>
}

export function Others ({ othersCount, expandFunc }) {
  if (othersCount < 0) return null

  const phrase = othersCount === 1 ? '1 other' : othersCount + ' others'

  return <React.Fragment>
    <span key='and'> and </span>
    <a key='others' styleName='communityLink' onClick={expandFunc}>{phrase}</a>
  </React.Fragment>
}
