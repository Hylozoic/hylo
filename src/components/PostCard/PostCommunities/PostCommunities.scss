.communities {
  padding-left: $space-4x;
  padding-right: 0;
  display: flex;
  justify-content: flex-start;
  border-top: 1px solid $color-cape-cod-10;
  overflow: hidden;
  min-height: 40px;
}

.constrained.communities {
  padding-left: 10px;
}

.communityList {
  composes: timestamp from 'css/typography.scss';
  overflow: hidden;
  text-overflow: ellipsis;
}

.communityLink {
  color: $color-caribbean-green;
  font-size: 12px;
  &:hover {
    text-decoration: none;
    color: $color-gossamer;
  }
}

.bottomBorder {
  border-bottom: 1px solid $color-cape-cod-10;
  margin-bottom: 10px;
}

.expanded {
  padding-top: 7px;
  width: 100%;
  padding-bottom: 10px;
  flex-wrap: wrap;
}

.row {
  display: flex;
  align-items: center;
  width: 100%;
}

.label {
  composes: timestamp from 'css/typography.scss';
}

.publicCommunityIcon {
  color: $color-caribbean-green;
  padding-left: 3px;
  font-size: 14px;
  top: 2px;
}

.expandLink {
  margin-left: auto;
  margin-right: 22px;
}

.expandIcon {
  composes: timestamp from 'css/typography.scss';
  font-size: 20px;
}

.expandedSection {
  width: 100%;
  padding-bottom: 10px;
}

.communityRow {
  display: flex;
  margin-bottom: 5px;
}

.communityCell {
  flex: 1;
  &:hover {
    text-decoration: none;
  }
}

.communityCellAvatar {
  @include image(20px);
  display: inline-block;
  vertical-align: middle;
  border-radius: 4px;
  margin-right: 10px;
}

.communityCellName {
  composes: text-button from 'css/typography.scss';
  font-size: 12px;
}

@media screen and (max-width: 325px) {

  .communityList {
    overflow: hidden;
    display: block;
    height: 20px;
    text-overflow: ellipsis;
    line-height: 20px;
    padding-left: 5px;
  }

  .label {
    text-align: right;
  }

}
