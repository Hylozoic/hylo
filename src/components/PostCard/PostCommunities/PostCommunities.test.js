import React from 'react'
import PostCommunities from './PostCommunities'
import { shallow } from 'enzyme'

describe('PostCommunities', () => {
  it('matches last snapshot', () => {
    const props = {
      communities: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        },
        {
          id: 2,
          name: 'Two',
          slug: 'two'
        },
        {
          id: 3,
          name: 'Three',
          slug: 'three'
        }
      ],
      slug: 'hylo'
    }

    const wrapper = shallow(<PostCommunities {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('matches last snapshot when expanded', () => {
    const props = {
      communities: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        },
        {
          id: 2,
          name: 'Two',
          slug: 'two'
        },
        {
          id: 3,
          name: 'Three',
          slug: 'three'
        }
      ],
      slug: 'hylo'
    }
    const wrapper = shallow(<PostCommunities {...props} />)
    wrapper.instance().toggleExpanded()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('returns null when in the only community', () => {
    const props = {
      communities: [
        {
          id: 1,
          name: 'One',
          slug: 'one'
        }
      ],
      slug: 'one'
    }

    const wrapper = shallow(<PostCommunities {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
