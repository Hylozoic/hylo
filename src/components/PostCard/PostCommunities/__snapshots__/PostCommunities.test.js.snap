// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCommunities matches last snapshot 1`] = `
<div
  data-stylename="communities"
>
  <div
    data-stylename="row"
  >
    <span
      data-stylename="label"
    >
      Posted In: 
    </span>
    <LinkedCommunityNameList
      communities={
        Array [
          Object {
            "id": 1,
            "name": "One",
            "slug": "one",
          },
          Object {
            "id": 2,
            "name": "Two",
            "slug": "two",
          },
          Object {
            "id": 3,
            "name": "Three",
            "slug": "three",
          },
        ]
      }
      expandFunc={[Function]}
      maxShown={2}
    />
    <a
      data-stylename="expandLink"
      onClick={[Function]}
    >
      <Icon
        data-stylename="expandIcon"
        name="ArrowDown"
      />
    </a>
  </div>
</div>
`;

exports[`PostCommunities matches last snapshot when expanded 1`] = `
<div
  data-stylename="communities expanded"
  onClick={[Function]}
>
  <div
    data-stylename="row"
  >
    <span
      data-stylename="label"
    >
      Posted In: 
    </span>
    <a
      data-stylename="expandLink"
      onClick={[Function]}
    >
      <Icon
        data-stylename="expandIcon"
        name="ArrowUp"
      />
    </a>
  </div>
  <CommunitiesList
    communities={
      Array [
        Object {
          "id": 1,
          "name": "One",
          "slug": "one",
        },
        Object {
          "id": 2,
          "name": "Two",
          "slug": "two",
        },
        Object {
          "id": 3,
          "name": "Three",
          "slug": "three",
        },
      ]
    }
  />
</div>
`;

exports[`PostCommunities returns null when in the only community 1`] = `""`;
